src/util/SweptCollection.js

```diff
@@ -0,0 +1,129 @@
+ * determining the entries lifetime.
+ * @property {boolean} [sweepArchivedOnly=false] Whether to only sweep entries with a property `archived: true`.
+ * E.g. for threads.
+ * @property {number} [sweepInterval=3600] How frequently, in seconds, to sweep the collection.
+ */
+
+/**
+ * A Collection which holds a max amount of entries and sweeps periodically.
+ * The first key is deleted if the Collection has reached max size.
+ * @extends {Collection}
+ * @param {SweptCollectionOptions|Iterable} [options=null] Options for constructing the swept collection.
+ */
+class SweptCollection extends Collection {
+  constructor(options = null) {
+    if (typeof options !== 'object') throw new TypeError('INVALID_TYPE', 'options', 'object or iterable', true);
+    const {
+      maxSize = -1,
+      maxSizePredicate = null,
+      sweepFunction = null,
+      sweepLifetime = 14400,
+      sweepLifetimeProperty = 'createdTimestamp',
+      sweepInterval = 3600,
+      sweepArchivedOnly = false,
+      client = null,
+      iterable = null,
+    } = options;
+
+    if (typeof maxSize !== 'number') throw new TypeError('INVALID_TYPE', 'maxSize', 'number');
+    if (maxSizePredicate !== null && typeof maxSizePredicate !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'maxSizePreidcate', 'function');
+    }
+    if (sweepFunction !== null && typeof sweepFunction !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'sweepFunction', 'function');
+    }
+    if (typeof sweepFunction !== 'function' && typeof sweepLifetime !== 'number') {
+      throw new TypeError('INVALID_TYPE', 'sweepLifetime', 'number');
+    }
+    if (typeof sweepFunction !== 'function' && typeof sweepLifetimeProperty !== 'string') {
+      throw new TypeError('INVALID_TYPE', 'sweepLifetimeProperty', 'number');
+    }
+    if (typeof sweepInterval !== 'number') throw new TypeError('INVALID_TYPE', 'sweepInterval', 'number');
+    if (typeof sweepFunction !== 'function' && typeof sweepArchivedOnly !== 'boolean') {
+      throw new TypeError('INVALID_TYPE', 'sweepArchivedOnly', 'boolean');
+    }
+    if (typeof client !== 'object') throw new TypeError('INVALID_TYPE', 'client', 'Client');
+
+    const definiteIterable = options[Symbol.iterator] ? options : iterable;
+    super(definiteIterable);
+
+    /**
+     * The client used to setInterval on, uses global setInterval if not provided.
+     * @type {?Client}
+     */
+    this.client = client;
+
+    /**
+     * The max size of the Collection.
+     * @type {number}
+     */
+    this.maxSize = maxSize;
+
+    /**
+     * A predicate for determining which item to remove when the Collection reaches max size.
+     * Using this can allow the size to grow larger than max size if none of the entries match the predicate.
+     * @type {?Function}
+     */
+    this.maxSizePredicate = maxSizePredicate;
+
+    const intervalFunction = sweepFunction
+      ? () => this.sweep(sweepFunction)
```

Comment:
This should be converted to a plain `function` since arrow functions don't have `this` and cannot be binded.

Reply:
I actually bound this out of habit, it's unnecessary. Arrow functions use the parent scope, so I'll just remove the bind